### DISCUSSION TITLE
Appropriate use of `AnyString` in creating a generic `Loaded<Entity, any>` type

### DISCUSSION THREAD

**Discussion by SpencerKaiser** [mikro-orm/mikro-orm#4192] [Q&A] [answered]
I have a method that accepts a variety of versions of an entity (e.g., `Loaded<Entity, 'user'> | Loaded<Entity, 'somethingElse'> | Loaded<Entity, 'anotherThing'>`) and I want to be able to pass any variation of those into this method... is this signature an appropriate use of `AnyString`?

```typescript
const myFunction = (entity: Entity | Loaded<Entity, AnyString>) => null;
```

The compiler seems happy, but just wanted to check to make sure I wasn't making use of it in the wrong way. I've been struggling with this for a while, so I'd be happy to update the docs with a little blurb about this if you think it'll add value.

### Edit
I lied... the compiler is not happy, so I'm unsure of how to accomplish this 😬

**Comment by B4nan** [comment]
I would go with `Loaded<T, any>`, that should do the job.

**Reply by SpencerKaiser** [reply]
That was my first guess, but the compiler isn't happy about that either...

```
error TS2345: Argument of type 'Loaded<InstantMeeting, "host" | "user">' is not assignable to parameter of type 'Loaded<InstantMeeting, any>'.
  Type 'Loaded<InstantMeeting, "host" | "user">' is not assignable to type 'InstantMeeting'.
    The types of 'toObject(...).user' are incompatible between these types.
      Type 'EntityDTO<User & {}> | undefined' is not assignable to type 'string | undefined'.
        Type 'EntityDTO<User & {}>' is not assignable to type 'string'.
```

Any idea??

**Reply by B4nan** [reply, marked as answer]
You could try v6, the types are quite refactored in there.

**Reply by SpencerKaiser** [reply]
I just tried and immediately got a ton of type errors, so I think I need to wait for the upgrade guide 😕

**Reply by SpencerKaiser** [reply]
Is there another type that has better overlap between an Entity and a Loaded Entity?

**Reply by SpencerKaiser** [reply]
Maybe I should just be passing the values in using `entity as Entity` when it's `Loaded<Entity, ...>`??

**Reply by B4nan** [reply]
There is an upgrading guide from the very first BC.

**Reply by SpencerKaiser** [reply]
Alright well I committed to the upgrade and now I have a ton of really confusing type errors related to circular references. I'm struggling to even read the errors because most have the dreaded `Type instantiation is excessively deep and possibly infinite.` associated with them ☹️ 

Any idea how to step through this?
```
src/api/bookings/post.ts(59,21): error TS2615: Type of property '$and' circularly references itself in mapped type '{ [K in keyof (OperatorMap<CallOption & { host: LoadedReference<Host & { user: LoadedReference<User & { email: Defined<string>; firstName: Defined<string>; lastName: Defined<string>; ... 18 more ...; setSchema: ((schema?: string | undefined) => void) & {}; }>; ... 24 more ...; setSchema: ((schema?: string | undefine...'.
```

**Reply by SpencerKaiser** [reply]
> There is an upgrading guide from the very first BC.

Oh and I forgot to address this... I didn't see it because it isn't in the sidebar in the upgrading section. I tried to figure out how that works to make a PR but couldn't figure out what was going on there.

**Reply by SpencerKaiser** [reply]
Alright I think I found what the issue was but could probably use some help making a repro if you're up for it... here's a single line that causes a type error:

``` typescript
const bookingOrNull = await em.findOne(Booking, {}, { populate: ['durationOption.callOption'] });
if (bookingOrNull) {
  const booking: Booking = bookingOrNull; // Throws type error below
}
```

```
error TS2589: Type instantiation is excessively deep and possibly infinite.
error TS2615: Type of property '$and' circularly references itself in mapped type '{ $and?: FilterQuery<DurationOption & { callOption: LoadedReference<CallOption & { host: LoadedReference<Host & { user: LoadedReference<User & { email: Defined<string>; firstName: Defined<...>; ... 19 more ...; setSchema: ((schema?: string | undefined) => void) & {}; }>; ... 24 more ...; setSchema: ((schema?: string...'.
error TS2615: Type of property '$and' circularly references itself in mapped type '{ [K in keyof (OperatorMap<CallOption & { host: LoadedReference<Host & { user: LoadedReference<User & { email: Defined<string>; firstName: Defined<string>; lastName: Defined<string>; ... 18 more ...; setSchema: ((schema?: string | undefined) => void) & {}; }>; ... 24 more ...; setSchema: ((schema?: string | undefine...'.
error TS2615: Type of property '$not' circularly references itself in mapped type '{ [K in keyof (OperatorMap<CallOption & { host: LoadedReference<Host & { user: LoadedReference<User & { email: Defined<string>; firstName: Defined<string>; lastName: Defined<string>; ... 18 more ...; setSchema: ((schema?: string | undefined) => void) & {}; }>; ... 24 more ...; setSchema: ((schema?: string | undefine...'.
[+30 more errors just like the above]
```

Removing the `.callOption` from the `populate` from that resolves the type error 😬 I was doing something similar and passing the result to a method that expected a `Booking` and it was throwing all those errors above 

## Edit
Found another example that does the same as the above with a different entity. I think the issue arises when this happens:
```
EntityA {
  b: EntityB
  c: EntityC
}

EntityB { ... }

EntityC { 
  d: EntityD
}

EntityD {
  b: EntityB
}
```

I think with a populate of C it pulls in the references to C.d.b which is the same as A.b... just a theory

**Reply by SpencerKaiser** [reply]
I'm a moron........... 


> ## TypeScript 4.9+ required

> Support for older TypeScript versions was dropped. 


I got it all working now; thanks for the help!

**Reply by B4nan** [reply]
So the v6 migration helped? That's great to hear!

**Reply by SpencerKaiser** [reply]
Yep! Some notable things on my end from the upgrade:
- I'm an idiot, but idiots need help sometimes... maybe throw a ⚠️ on the TypeScript upgrade header?
  - Part of this was because I saw "**_older_** TypeScript versions" and kinda misread as "old" and through to myself "**I'm on a pretty current version, this doesn't apply to me**" 🙃 
- I have been seriously misusing `.load` for quite a while... I was using that in conjunction with `unwrap` when it really sounds like I should have been saving the result of `.load` and using that
  - However, this becomes a bit of an issue when passing that entity to a method... BUT I've just been passing it and using `.load` (which adds some safety in case I remove it in the parent and because it won't requery if it's there)
  - It might be worth making a few verbiage changes and/or adding example to make this clearer to users... it sounds like (from our other convo) that it isn't straight forward to have `.load` modify the type of the parent
    - Happy to contribute this if you think it'd add value!
- Some of the changes around `.toPOJO` were a bit confusing, but I don't totally remember what the context was... pretty sure it was related to my use of `Loaded` (see below) because now all my uses of `toPOJO` are solid
- I used `Loaded` for a few things, but primarily 2 use cases: to give method args a type and as the type for a `let` where the assignment was handled inside a `try...catch` that was explicitly wrapped around a DB fetch (because we didn't want the catch to be super ambiguous)
  - After this change, the method piece is totally irrelevant... I just pass the entity as the entity and then use redundant `.load` calls (like I mentioned above) if I need a nested attribute inside that method 👌 

Overall v6 is a MASSIVE improvement... the types are way cleaner (both in terms of use and in terms of the tooltip in VSCode) and it seems like build time is much, much faster. Awesome work (as always) and thanks for the help! Lemme know if you'd like me to contribute any of the things I mentioned above.

**Reply by SpencerKaiser** [reply]
Actually I am seeing more of the circular reference errors... swapping the entity attributes that are causing the circular reference to being `ref: true` + `Ref<Entity>` resolves the issue... I'm gonna try and see if I can make a repro. Do you have a good StackBlitz repro template?

**Reply by B4nan** [reply]
You can try `Rel`, which is just an identity type.

https://stackblitz.com/edit/mikro-orm-esm-repro

**Reply by B4nan** [reply]
Or maybe base it on this one, stackblitz has outdated sqlite which doesnt support returning statements, and this example has that worked around by patching some internals.

https://stackblitz.com/edit/mikro-orm-v6-strict-partial-loading

**Reply by B4nan** [reply]
> It might be worth making a few verbiage changes and/or adding example to make this clearer to users... it sounds like (from our other convo) that it isn't straight forward to have .load modify the type of the parent

Any docs improvements are more than welcome. Unfortunately TS does not support async type guards, which is the only way to modify type of existing properties, but technically `load` modifies the state of the wrapped entity.

**Reply by SpencerKaiser** [reply]
Well that was exceptionally painful to track down 😅 looks like the whole thing originates from circular references (unsurprisingly) when used with `BaseEntity`.

Take a look at [the repro](https://stackblitz.com/edit/mikro-orm-v6-strict-partial-loading-tm64dy?file=typeError.ts) and let me know what you think!

**Reply by B4nan** [reply]
https://github.com/mikro-orm/mikro-orm/commit/0ca77f3a49e0c10de95b746bc2c3cc7529245532

**Reply by SpencerKaiser** [reply]
You're a legend 🍻
